File: core/core-api/src/main/java/com/coffee/gu/api/controller/v1/response/IssuedCouponResponse.java
package com.coffee.gu.api.controller.v1.response;

import com.coffee.gu.coupon.IssuedCoupon;
import com.coffee.gu.enums.CouponType;
import com.coffee.gu.enums.IssuedCouponState;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

public record IssuedCouponResponse(
        Long id,
        IssuedCouponState state,
        String name,
        CouponType type,
        BigDecimal discount,
        LocalDateTime expiredAt
) {
    public static IssuedCouponResponse from(IssuedCoupon issuedCoupon) {
        return new IssuedCouponResponse(
                issuedCoupon.getId(),
                issuedCoupon.getState(),
                issuedCoupon.getCoupon().getName(),
                issuedCoupon.getCoupon().getType(),
                issuedCoupon.getCoupon().getDiscount(),
                issuedCoupon.getCoupon().getExpiredAt()
        );
    }

    public static List<IssuedCouponResponse> from(List<IssuedCoupon> issuedCoupons) {
        return issuedCoupons.stream()
                .map(IssuedCouponResponse::from)
                .toList();
    }
}

File: core/core-api/src/main/kotlin/com/coffee/gu/api/controller/v1/response/IssuedCouponResponse.kt
package com.coffee.gu.api.controller.v1.response

import com.coffee.gu.coupon.IssuedCoupon
import com.coffee.gu.enums.CouponType
import com.coffee.gu.enums.IssuedCouponState
import java.math.BigDecimal
import java.time.LocalDateTime

class IssuedCouponResponse(
    val id: Long,
    val state: IssuedCouponState,
    val name: String,
    val type: CouponType,
    val discount: BigDecimal,
    val expiredAt: LocalDateTime,
) {
    companion object {
        @JvmStatic
        fun from(issuedCoupon: IssuedCoupon): IssuedCouponResponse {
            return IssuedCouponResponse(
                id = issuedCoupon.id,
                state = issuedCoupon.state,
                name = issuedCoupon.coupon.name,
                type = issuedCoupon.coupon.type,
                discount = issuedCoupon.coupon.discount,
                expiredAt = issuedCoupon.coupon.expiredAt
            )
        }

        @JvmStatic
        fun from(issuedCoupons: List<IssuedCoupon>): List<IssuedCouponResponse> {
            return issuedCoupons.map { from(it) }
        }
    }
}
